casper/src/main/scala/coop/rchain/casper/blocks/BlockReceiver.scala

```diff
@@ -0,0 +1,295 @@
+package coop.rchain.casper.blocks
+
+import cats.Show
+import cats.effect.concurrent.Ref
+import cats.effect.{Concurrent, Sync}
+import cats.syntax.all._
+import coop.rchain.blockstorage.BlockStore
+import coop.rchain.blockstorage.BlockStore.BlockStore
+import coop.rchain.blockstorage.dag.BlockDagStorage
+import coop.rchain.casper.engine.BlockRetriever
+import coop.rchain.casper.protocol.BlockMessage
+import coop.rchain.casper.syntax._
+import coop.rchain.casper.util.ProtoUtil
+import coop.rchain.casper.{CasperShardConf, PrettyPrinter, Validate}
+import coop.rchain.models.BlockHash.BlockHash
+import coop.rchain.shared.Log
+import coop.rchain.shared.syntax._
+import fs2.Stream
+import fs2.concurrent.Queue
+
+sealed trait RecvStatus
+// Begin checking and storing block
+case object BeginStoreBlock extends RecvStatus
+// Block stored in the block store, waiting for validation and DAG insertion
+case object EndStoreBlock extends RecvStatus
+// Block sent to validation
+case object PendingValidation extends RecvStatus
+// Requested missing dependencies
+case object Requested extends RecvStatus
+
+object BlockReceiverState {
+  def apply[MId: Show]: BlockReceiverState[MId] = BlockReceiverState(
+    blocksSt = Map(),
+    receiveSt = Map(),
+    childRelations = Map()
+  )
+}
+
+/**
+  * Block receiver state
+  *
+  * It consist of three events. Two to store blocks (begin and end) to prevent race when
+  * storing blocks and finished when block is validated and added to the DAG (end of processing).
+  */
+final case class BlockReceiverState[MId: Show](
```

Comment:
It would be great to have some unit tests for this case class.

Reply:
You are right, it should be added in another PR. I created an issue for that as mentioned in the description.
- #3723